homeassistant/components/sensor/modem_callerid.py

```diff
@@ -0,0 +1,125 @@
+    vol.Optional(CONF_NAME, default=DEFAULT_NAME): cv.string,
+    vol.Optional(CONF_DEVICE, default=DEFAULT_DEVICE): cv.string
+})
+
+
+def setup_platform(hass, config, add_devices, discovery_info=None):
+    """Setup modem caller id sensor platform."""
+    name = config.get(CONF_NAME)
+    port = config.get(CONF_DEVICE)
+    add_devices([ModemCalleridSensor(hass, name, port)])
+
+
+class ModemCalleridSensor(Entity):
+    """Implementation of USB modem callerid sensor."""
+
+    def __init__(self, hass, name, port):
+        """Initialize the sensor."""
+        from basicmodem.basicmodem import BasicModem as bm
+        self._state = STATE_UNAVAILABLE
+        self._attributes = {"cid_time": 0, "cid_number": '', "cid_name": ''}
+        self._name = name
+        self.port = port
+        self.modem = None
+
+        self.modem = bm(self.port, self._incomingcallcallback)
+        if self.modem.state == bm.STATE_FAILED:
+            _LOGGER.error('Unable to initialize modem.')
+            return
+
+        self.set_state(STATE_IDLE)
+        hass.bus.listen_once(EVENT_HOMEASSISTANT_STOP, self._stop_modem)
+
+    def set_state(self, state):
+        """Set the state."""
+        self._state = state
+
+    def set_attributes(self, attributes):
+        """Set the state attributes."""
+        self._attributes = attributes
+
+    @property
+    def should_poll(self):
+        """No polling needed."""
+        return False
+
+    @property
+    def icon(self):
+        """Return icon."""
+        return ICON
+
+    @property
+    def state(self):
+        """Return the state of the device."""
+        return self._state
+
+    @property
+    def name(self):
+        """Return the name of the sensor."""
+        return self._name
+
+    @property
+    def device_state_attributes(self):
+        """Return the state attributes."""
+        return self._attributes
+
+    def _stop_modem(self, event):
+        """HA is shutting down, close modem port."""
+        if self.modem:
+            self.modem.close()
+            self.modem = None
+        return
+
+    def _incomingcallcallback(self, bm, newstate):
+        """Callback from modem, process based on new state."""
+        from basicmodem.basicmodem import BasicModem as bm
+        if newstate == bm.STATE_RING:
+            if self.state == bm.STATE_IDLE:
+                att = {"cid_time": bm.get_cidtime,
+                       "cid_number": '',
+                       "cid_name": ''}
+                self.set_attributes(att)
+
+            self.set_state(STATE_RING)
+            self.update_ha_state()
+        elif newstate == bm.STATE_CALLERID:
+            att = {"cid_time": bm.get_cidtime,
+                   "cid_number": bm.get_cidnumber,
+                   "cid_name": bm.get_cidname}
+            self.set_attributes(att)
+            self.set_state(STATE_CALLERID)
+            self.update_ha_state()
+        elif newstate == bm.STATE_IDLE:
+            self.set_state(STATE_IDLE)
```

Comment:
No reason to create a `set_state` method if you're already in the same class. You could use do `self._state = STATE_IDLE`